source/slang/slang-capability.h

```diff
@@ -189,62 +132,177 @@ struct CapabilitySet
 
     /// Does this capability set imply all the capabilities in `other`?
     bool implies(CapabilitySet const& other) const;
 
-    /// Does this capability set imply all the capabilities in `other`?
-    bool implies(CapabilityConjunctionSet const& other) const;
-
     /// Does this capability set imply the atomic capability `other`?
     bool implies(CapabilityAtom other) const;
 
-    /// Join two capability sets to form (this & other).
+    /// Join two capability sets to form ('this' & 'other').
+    /// Destroy incompatible targets/sets apart of 'this' between ('this' & 'other').
+    /// `this` may be made invalid if other is fully disjoint.
     void join(const CapabilitySet& other);
 
-    void unionWith(const CapabilityConjunctionSet& other);
-
-    void simpleJoinWithSetMask(const CapabilitySet& other, CapabilityName abstractMask);
+    /// Join two capability sets to form ('this' & 'other'). 
+    /// If a target/set has an incompatable atom, do not destroy the target/set.
+    void nonDestructiveJoin(CapabilitySet& other);
 
-    CapabilitySet getTargetsThisIsMissingFromOther(const CapabilitySet& other);
+    /// Add all targets/sets of 'other' into 'this'. Overlapping sets are removed.
+    void unionWith(const CapabilitySet& other);
 
-    void canonicalize();
+    /// Return a capability set of 'target' atoms 'this' has, but 'other' does not. 
+    CapabilitySet getTargetsThisHasButOtherDoesNot(const CapabilitySet& other);
 
         /// Are these two capability sets equal?
     bool operator==(CapabilitySet const& that) const;
 
-    /// Get access to the raw atomic capabilities that define this set.
-    List<CapabilityConjunctionSet>& getExpandedAtoms() { return m_conjunctions; }
-    const List<CapabilityConjunctionSet>& getExpandedAtoms() const { return m_conjunctions; }
-
-
+    void addCapability(List<List<CapabilityAtom>>& atomLists);
     /// Calculate a list of "compacted" atoms, which excludes any atoms from the expanded list that are implies by another item in the list.
-    void calcCompactedAtoms(List<List<CapabilityAtom>>& outAtoms) const;
-
-    bool isBetterForTarget(CapabilitySet const& that, CapabilitySet const& targetCaps) const;
-
-    static bool checkCapabilityRequirement(CapabilitySet const& available, CapabilitySet const& required, const CapabilityConjunctionSet*& outFailedAvailableSet);
-
-    bool isExactSubset(CapabilitySet const& maybeSuperSet);
 
+    bool isBetterForTarget(CapabilitySet const& that, CapabilitySet const& targetCaps, bool& isEqual) const;
+
+    /// Find any capability sets which are in 'available' but not in 'required'. Return false if this situation occurs. 
+    static bool checkCapabilityRequirement(CapabilitySet const& available, CapabilitySet const& required, CapabilityAtomSet& outFailedAvailableSet);
+
+    inline void addToTargetCapabilityWithValidUIntSetAndTargetAndStage(const CapabilityName& target, const CapabilityName& stage, const CapabilityAtomSet& setToAdd);
```

Comment:
For methods that may take and move the parameters, consider just pass by value to allow arguments to be moved straight through. Pass by reference actually prevents copy ellision.